<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Alignment tools &mdash; wgd 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="wgd 1.0.0 documentation" href="index.html" />
    <link rel="next" title="Ks distribution module" href="ks.html" />
    <link rel="prev" title="Codeml python wrapper" href="codeml.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="ks.html" title="Ks distribution module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="codeml.html" title="Codeml python wrapper"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-wgd.alignment">
<span id="alignment-tools"></span><h1>Alignment tools<a class="headerlink" href="#module-wgd.alignment" title="Permalink to this headline">¶</a></h1>
<p>Arthur Zwaenepoel - 2017</p>
<p>Alignment related tools.</p>
<p>Contribution note: It is trivial to add support for another aligner (besides MUSCLE and PRANK).
Especially when it&#8217;s for amino acid alignments (codon alignment as PRANK are a bit more involved).
To add another aligner, only the <a class="reference internal" href="#wgd.alignment.MSA.run_aligner" title="wgd.alignment.MSA.run_aligner"><code class="xref py py-meth docutils literal"><span class="pre">MSA.run_aligner()</span></code></a> method should be modified to include
the aligner of interest.</p>
<dl class="class">
<dt id="wgd.alignment.MSA">
<em class="property">class </em><code class="descclassname">wgd.alignment.</code><code class="descname">MSA</code><span class="sig-paren">(</span><em>muscle='muscle'</em>, <em>prank='prank'</em>, <em>tmp='./'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment.MSA" title="Permalink to this definition">¶</a></dt>
<dd><p>Multiple multiple sequence alignment (MSA) programs python wrappers.
Currently only runs with default settings.</p>
<p>Usage examples</p>
<ul>
<li><p class="first">Sequences in dictionary, output as dictionary:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">MSA</span><span class="p">()</span><span class="o">.</span><span class="n">run_aligner</span><span class="p">({</span><span class="s1">&#39;bear_gene&#39;</span><span class="p">:</span> <span class="s1">&#39;BEARBEAR&#39;</span><span class="p">,</span> <span class="s1">&#39;hare_gene&#39;</span><span class="p">:</span> <span class="s1">&#39;HAREERAH&#39;</span><span class="p">,</span> <span class="s1">&#39;yeast_gene&#39;</span><span class="p">:</span> <span class="s1">&#39;BEERBEARBEER&#39;</span><span class="p">})</span>
<span class="go">{&#39;bear_gene&#39;: &#39;----BEARBEAR-&#39;,</span>
<span class="go">&#39;hare_gene&#39;: &#39;-----HAREERAH&#39;,</span>
<span class="go">&#39;yeast_gene&#39;: &#39;BEERBEARBEER-&#39;}</span>
</pre></div>
</div>
</li>
<li><p class="first">Sequences in fasta file, output as fasta file <code class="docutils literal"><span class="pre">(msa.fasta)</span></code>:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">MSA</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span><span class="o">.</span><span class="n">run_aligner</span><span class="p">(</span><span class="s1">&#39;./sequences.fasta&#39;</span><span class="p">,</span> <span class="s1">&#39;./msa.fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
<dl class="method">
<dt id="wgd.alignment.MSA.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>muscle='muscle'</em>, <em>prank='prank'</em>, <em>tmp='./'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment.MSA.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Muscle wrapper init.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>muscle</strong> &#8211; path to Muscle executable, will by defult look for Muscle in the system PATH</li>
<li><strong>prank</strong> &#8211; path to PRANK executable</li>
<li><strong>tmp</strong> &#8211; directory to store temporary files.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="wgd.alignment.MSA.run_aligner">
<code class="descname">run_aligner</code><span class="sig-paren">(</span><em>sequences</em>, <em>aligner='muscle'</em>, <em>file=None</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment.MSA.run_aligner" title="Permalink to this definition">¶</a></dt>
<dd><p>Run MSA on sequences stored in dictionary with a particular aligner</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sequences</strong> &#8211; dictionary of sequences or input fasta file</li>
<li><strong>file</strong> &#8211; output file name, if not provided output is dictionary</li>
<li><strong>aligner</strong> &#8211; alignment software to use (prank|muscle)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">MSA file</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="wgd.alignment._nucleotide_msa_from_protein_msa">
<code class="descclassname">wgd.alignment.</code><code class="descname">_nucleotide_msa_from_protein_msa</code><span class="sig-paren">(</span><em>protein_msa_sequences</em>, <em>nucleotide_sequences</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment._nucleotide_msa_from_protein_msa" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a nucleotide multiple sequence alignment from a protein multiple sequence alignment</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>protein_msa_sequences</strong> &#8211; dictionary with protein sequences (aligned)</li>
<li><strong>nucleotide_sequences</strong> &#8211; dictionary with nucleotide sequences (unaligned)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">dictionary with nucleotide sequences (aligned)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.alignment._strip_gaps">
<code class="descclassname">wgd.alignment.</code><code class="descname">_strip_gaps</code><span class="sig-paren">(</span><em>msa_dict</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment._strip_gaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Strip gap positions from a multiple sequence alignment (MSA).
Finds the positions in the strings that have a gap and removes them in all sequences.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>msa_dict</strong> &#8211; dictionary with aligned nucleotide sequences.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.alignment.hamming_distance">
<code class="descclassname">wgd.alignment.</code><code class="descname">hamming_distance</code><span class="sig-paren">(</span><em>s1</em>, <em>s2</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment.hamming_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the Hamming distance between equal-length sequences</p>
</dd></dl>

<dl class="function">
<dt id="wgd.alignment.multiple_sequence_aligment_nucleotide">
<code class="descclassname">wgd.alignment.</code><code class="descname">multiple_sequence_aligment_nucleotide</code><span class="sig-paren">(</span><em>msa_protein</em>, <em>nucleotide_sequences</em>, <em>min_length=100</em>, <em>aligner='muscle'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment.multiple_sequence_aligment_nucleotide" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a nucleotide multiple sequence alignment based on a protein alignment</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>msa_protein</strong> &#8211; dictionary of aligned protein sequences</li>
<li><strong>nucleotide_sequences</strong> &#8211; nucleotide sequence dictionary</li>
<li><strong>min_length</strong> &#8211; minimum alignment length to consider</li>
<li><strong>aligner</strong> &#8211; alignment program used, if prank, than msa_protein will be interpreted as codon alignment</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">nucleotide MSA, length, stripped length</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.alignment.pairwise_alignment_stats">
<code class="descclassname">wgd.alignment.</code><code class="descname">pairwise_alignment_stats</code><span class="sig-paren">(</span><em>msa</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.alignment.pairwise_alignment_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Get pairwise stats from MSA.
Return as  dictionary {Gene1: {Gene2: [0.8, 0.7], Gene3: [...], ...}</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>msa</strong> &#8211; MSA dictionary</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dictionary</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="codeml.html"
                        title="previous chapter">Codeml python wrapper</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="ks.html"
                        title="next chapter">Ks distribution module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/alignment.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="ks.html" title="Ks distribution module"
             >next</a> |</li>
        <li class="right" >
          <a href="codeml.html" title="Codeml python wrapper"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Arthur Zwaenepoel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.6.
    </div>
  </body>
</html>