<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Command line interface &mdash; wgd 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="wgd 0.0.1 documentation" href="index.html" />
    <link rel="next" title="wgd_cli" href="wgd_cli.html" />
    <link rel="prev" title="wgd: Whole genome duplication analysis in Python" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="wgd_cli.html" title="wgd_cli"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="wgd: Whole genome duplication analysis in Python"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="command-line-interface">
<span id="cli"></span><h1>Command line interface<a class="headerlink" href="#command-line-interface" title="Permalink to this headline">¶</a></h1>
<p>The command line interface (CLI) and all command line utilities are written in python3.5 using the click library.
Upon installation, all available command line utilities bundled in the CLI can be listed by running:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>$ wgd --help

Usage: wgd [OPTIONS] COMMAND [ARGS]...

  Welcome to the wgd command line interface!

                         _______
                         \  ___ `&#39;.
         _     _ .--./)   &#39; |--.\  \
   /\    \\   ///.&#39;&#39;\\    | |    \  &#39;
   `\\  //\\ //| |  | |   | |     |  &#39;
     \`//  \&#39;/  \`-&#39; /    | |     |  |
      \|   |/   /(&quot;&#39;`     | |     &#39; .&#39;
       &#39;        \ &#39;---.   | |___.&#39; /&#39;
                 /&#39;&quot;&quot;&#39;.\ /_______.&#39;/
                ||     ||\_______|/
                \&#39;. __//
                 `&#39;---&#39;

  Arthur Zwaenepoel - 2017

Options:
  --verbose [silent|info|debug]  Verbosity level, default = info.
  -h, --help                     Show this message and exit.

Commands:
  blast       All-vs.-all blastp (+ MCL) analysis.
  hist        Plot (stacked) histograms.
  ks          Ks distribution construction.
  mix         Mixture modeling of Ks distributions.
  pipeline_1  Standard workflow whole paranome Ks.
  pipeline_2  Standard workflow one-vs-one ortholog Ks.
  syn         Co-linearity analyses.
</pre></div>
</div>
<p>Note that the verbose flag can be set before the subcommands:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>$ wgd --verbose [silent|info|debug] [COMMAND]
</pre></div>
</div>
<p>Which sets the verbosity for the logging.</p>
</div>
<div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<div class="section" id="whole-paranome-ks-distribution-construction">
<h2>Whole paranome K<sub>S</sub> distribution construction<a class="headerlink" href="#whole-paranome-ks-distribution-construction" title="Permalink to this headline">¶</a></h2>
<p>Here a recipe is shown for K<sub>S</sub> distribution construction using the provided commands in the <code class="docutils literal"><span class="pre">wgd</span></code>
command line interface. For this analysis we could also use the provided pipeline <code class="docutils literal"><span class="pre">wgd</span> <span class="pre">pipeline_1</span></code>
command, which provides less control, but is quite convenient in usage for this standard workflow.</p>
<p><strong>1) Full control</strong></p>
<p>First we infer paralogous families using all-vs-all <cite>blastp</cite> and <cite>mcl</cite>:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>$ wgd blast --cds --mcl -s penguin.cds.fasta -o ./
</pre></div>
</div>
<p>Then we can construct a K<sub>S</sub> distribution for the paranome we just generated
in the file <code class="docutils literal"><span class="pre">penguin.cds.fasta.mcl</span></code> using <cite>wgd ks</cite>:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>$ wgd ks -gf penguin.cds.fasta.mcl -s penguin.cds.fasta -o ./
</pre></div>
</div>
<p>If we have adequate structural annotation data available, we can look for anchor
pairs and make a K<sub>S</sub> distribution for the obtained co-linear
gene pairs:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>$ wgd coll -ks penguin.cds.fasta.ks.tsv -gff annotation.gff -gf penguin.cds.fasta.mcl -o ./
</pre></div>
</div>
<p>That&#8217;s it! Quite easy no?</p>
<p><strong>2) Pipeline</strong></p>
<p>The same can be accomplished using the <code class="docutils literal"><span class="pre">pipeline_1</span></code> command:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>$ wgd pipeline_1 -gff annotation.gff penguin.cds.fasta penguin.wgd_out
</pre></div>
</div>
<p>However this dos not provide any further control, as opposed to the subcommands listed above.
Note that this ability to for more detailed control was not illustrated above.</p>
<p>More detailed documentation of the main <code class="docutils literal"><span class="pre">wgd</span></code> commands can be obtained here:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="wgd_cli.html">wgd_cli</a></li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Command line interface</a></li>
<li><a class="reference internal" href="#examples">Examples</a><ul>
<li><a class="reference internal" href="#whole-paranome-ks-distribution-construction">Whole paranome K<sub>S</sub> distribution construction</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">wgd: Whole genome duplication analysis in Python</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="wgd_cli.html"
                        title="next chapter">wgd_cli</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/cli.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="wgd_cli.html" title="wgd_cli"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="wgd: Whole genome duplication analysis in Python"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Arthur Zwaenepoel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.6.
    </div>
  </body>
</html>