<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>wgd: Whole genome duplication analysis in Python &mdash; wgd 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="wgd 1.0.0 documentation" href="#" />
    <link rel="next" title="Command line interface" href="cli.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cli.html" title="Command line interface"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">wgd 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="wgd-whole-genome-duplication-analysis-in-python">
<h1>wgd: Whole genome duplication analysis in Python<a class="headerlink" href="#wgd-whole-genome-duplication-analysis-in-python" title="Permalink to this headline">¶</a></h1>
<p>This Python package and corresponding command line interface (CLI) were developed for various analyses related
to whole genome duplications (WGDs). Here the Python API is documented as well as the various command line
utilities bundled in the <code class="docutils literal"><span class="pre">wgd</span></code> CLI.</p>
<p>To install <code class="docutils literal"><span class="pre">wgd</span></code>, clone the repository available at <a class="reference external" href="https://github.com/arzwa/wgd">https://github.com/arzwa/wgd</a>, navigate into the repo, run
<code class="docutils literal"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">.</span></code>.</p>
<div class="section" id="external-software">
<h2>External software<a class="headerlink" href="#external-software" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">wgd</span></code> requires the following third party executables (preferably these should also be in the <cite>PATH</cite> environment variable):</p>
<p>For <code class="docutils literal"><span class="pre">wgd</span> <span class="pre">blast</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">blast</span></code> (Altschul _et al._ 2008), from which it uses the <code class="docutils literal"><span class="pre">blastp</span></code> and <code class="docutils literal"><span class="pre">makeblastdb</span></code> commands, <code class="docutils literal"><span class="pre">sudo</span> <span class="pre">apt-get</span> <span class="pre">install</span> <span class="pre">ncbi-blast+</span></code> will often suffice for installation</li>
<li><code class="docutils literal"><span class="pre">mcl</span></code> (<a class="reference external" href="https://micans.org/mcl/index.html">https://micans.org/mcl/index.html</a>). Get MCL using your package manager <code class="docutils literal"><span class="pre">sudo</span> <span class="pre">apt-get</span> <span class="pre">install</span> <span class="pre">mcl</span></code> or download it at the provided link.</li>
</ul>
<p>For <code class="docutils literal"><span class="pre">wgd</span> <span class="pre">ks</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">muscle</span></code> (Edgar 2004) for multiple sequence alignment. MUSCLE can be retrieved from <a class="reference external" href="http://www.drive5.com/muscle/">http://www.drive5.com/muscle/</a>. Codon alignment with PRANK (Löytynoja &amp; Goldman 2005, 2008)  is also supported (<a class="reference external" href="http://wasabiapp.org/software/prank/">http://wasabiapp.org/software/prank/</a>).</li>
<li><code class="docutils literal"><span class="pre">codeml</span></code> from the PAML software package (Yang 1997). PAML can be downloaded from the following link: <a class="reference external" href="http://abacus.gene.ucl.ac.uk/software/paml.html">http://abacus.gene.ucl.ac.uk/software/paml.html</a></li>
<li>Depending on which weighting method you choose for node-weighting the Ks distribution, you might also need <code class="docutils literal"><span class="pre">FastTree</span></code> (<a class="reference external" href="http://www.microbesonline.org/fasttree/">http://www.microbesonline.org/fasttree/</a>) (Price <cite>et al.</cite> 2010) or <code class="docutils literal"><span class="pre">phyml</span></code> (<a class="reference external" href="http://www.atgc-montpellier.fr/phyml/">http://www.atgc-montpellier.fr/phyml/</a>) (Guindon &amp; Gascuel 2003) (not necessary however).</li>
</ul>
<p>For <code class="docutils literal"><span class="pre">wgd</span> <span class="pre">syn</span></code></p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">i-adhore</span></code> from the I-ADHoRe 3.0 suite (<a class="reference external" href="http://bioinformatics.psb.ugent.be/beg/tools/i-adhore30">http://bioinformatics.psb.ugent.be/beg/tools/i-adhore30</a>) (Proost <cite>et al.</cite> 2012)</li>
</ul>
<p>Of course, you don&#8217;t need all of these tools installed for each feature of <code class="docutils literal"><span class="pre">wgd</span></code>.</p>
</div>
<div class="section" id="command-line-tools">
<h2>Command line tools<a class="headerlink" href="#command-line-tools" title="Permalink to this headline">¶</a></h2>
<p>The command line tools are the main functionality of the <code class="docutils literal"><span class="pre">wgd</span></code> package. You can find tools for the following
analyses:</p>
<ol class="arabic simple">
<li>All-<em>versus</em>-all Blastp analysis and MCL clustering</li>
<li>Whole paranome K<sub>S</sub> (and K<sub>A</sub> and ω) distribution construction</li>
<li>One-versus-one orthologs K<sub>S</sub> (and K<sub>A</sub> and ω;) distribution construction</li>
<li>Mixture modeling of K<sub>S</sub> distributions and WGD-specific paralog extraction</li>
<li>Interactive visualization of (multiple) K<sub>S</sub> distributions and kernel density estimates thereof</li>
<li>Intragenomic co-linearity/synteny analysis and anchor based K<sub>S</sub> distribution construction</li>
<li>Co-linearity dotplot construction</li>
</ol>
<p>All relative information can be find here:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html#examples">Examples</a></li>
</ul>
</div>
</div>
<div class="section" id="python-package">
<h2>Python package<a class="headerlink" href="#python-package" title="Permalink to this headline">¶</a></h2>
<p>For those interested in the underlying structure of <code class="docutils literal"><span class="pre">wgd</span></code>, here you can find the full documentation of the API.</p>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="blast_mcl.html">Blast &amp; Markov clustering (MCL)</a></li>
<li class="toctree-l1"><a class="reference internal" href="codeml.html">Codeml anlysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="alignment.html">Alignment tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="ks.html">Ks distribution analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="phy.html">Phylogenetics tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="mix.html">Mixture modeling</a></li>
<li class="toctree-l1"><a class="reference internal" href="syn.html">Co-linearity analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">Utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="viz.html">Visualization</a></li>
</ul>
</div>
</div>
<div class="section" id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span>Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span>Search Page</span></a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">wgd: Whole genome duplication analysis in Python</a><ul>
<li><a class="reference internal" href="#external-software">External software</a></li>
<li><a class="reference internal" href="#command-line-tools">Command line tools</a></li>
<li><a class="reference internal" href="#python-package">Python package</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href="cli.html"
                        title="next chapter">Command line interface</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cli.html" title="Command line interface"
             >next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">wgd 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Arthur Zwaenepoel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.6.
    </div>
  </body>
</html>