<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Ks distribution module &mdash; wgd 0.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="wgd 0.0.1 documentation" href="index.html" />
    <link rel="next" title="Mixture modeling" href="mix.html" />
    <link rel="prev" title="Alignment tools" href="alignment.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mix.html" title="Mixture modeling"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="alignment.html" title="Alignment tools"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-wgd.ks_distribution">
<span id="ks-distribution-module"></span><h1>Ks distribution module<a class="headerlink" href="#module-wgd.ks_distribution" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="wgd.ks_distribution._average_linkage_clustering">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">_average_linkage_clustering</code><span class="sig-paren">(</span><em>pairwise_estimates</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution._average_linkage_clustering" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform average linkage clustering using fastcluster.
The first two columns of the output contain the node indices which are joined in each step.
The input nodes are labeled 0, . . . , N - 1, and the newly generated nodes have the labels N, . . . , 2N - 2.
The third column contains the distance between the two nodes at each step, ie. the
current minimal distance at the time of the merge. The fourth column counts the
number of points which comprise each new node.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pairwise_estimates</strong> &#8211; dictionary with data frames with pairwise estimates of Ks, Ka and Ka/Ks
(or at least Ks), as returned by <a class="reference internal" href="#wgd.ks_distribution.analyse_family" title="wgd.ks_distribution.analyse_family"><code class="xref py py-func docutils literal"><span class="pre">analyse_family()</span></code></a>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">average linkage clustering as performed with <code class="docutils literal"><span class="pre">fastcluster.average</span></code>.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution._calculate_weighted_ks">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">_calculate_weighted_ks</code><span class="sig-paren">(</span><em>clustering</em>, <em>pairwise_estimates</em>, <em>family_id=None</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution._calculate_weighted_ks" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate weighted Ks, Kn and w values following the procedure as outlined in Vanneste et al. (2013)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>clustering</strong> &#8211; average linkage clustering results as produced with <a class="reference internal" href="#wgd.ks_distribution._average_linkage_clustering" title="wgd.ks_distribution._average_linkage_clustering"><code class="xref py py-func docutils literal"><span class="pre">_average_linkage_clustering()</span></code></a></li>
<li><strong>pairwise_estimates</strong> &#8211; pairwise Ks, Kn and w estimates as produced by <code class="xref py py-meth docutils literal"><span class="pre">codeml.Codeml.run_codeml()</span></code></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a pandas data frame of weighted Ks, Kn and w values.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.analyse_family">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">analyse_family</code><span class="sig-paren">(</span><em>family_id</em>, <em>family</em>, <em>nucleotide</em>, <em>tmp='./'</em>, <em>muscle='muscle'</em>, <em>codeml='codeml'</em>, <em>preserve=False</em>, <em>times=1</em>, <em>min_length=100</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.analyse_family" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper function for the analysis of one paralog family. Performs alignment with
<code class="xref py py-meth docutils literal"><span class="pre">alignment.Muscle.run_muscle()</span></code> and codeml analysis with <code class="xref py py-meth docutils literal"><span class="pre">codeml.Codeml.run_codeml()</span></code>.
Subsequently also clustering with <a class="reference internal" href="#wgd.ks_distribution._average_linkage_clustering" title="wgd.ks_distribution._average_linkage_clustering"><code class="xref py py-func docutils literal"><span class="pre">_average_linkage_clustering()</span></code></a> is performed and weighted Ks, Kn and w
values are calculated using <a class="reference internal" href="#wgd.ks_distribution._calculate_weighted_ks" title="wgd.ks_distribution._calculate_weighted_ks"><code class="xref py py-func docutils literal"><span class="pre">_calculate_weighted_ks()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>family_id</strong> &#8211; gene family id</li>
<li><strong>family</strong> &#8211; dictionary with sequences of paralogs</li>
<li><strong>nucleotide</strong> &#8211; nucleotide (CDS) sequences dictionary</li>
<li><strong>tmp</strong> &#8211; tmp directory</li>
<li><strong>muscle</strong> &#8211; muscle path</li>
<li><strong>codeml</strong> &#8211; codeml path</li>
<li><strong>preserve</strong> &#8211; preserve intermediate files</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">csv</span></code> file with results for the paralog family of interest</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.analyse_family_ortholog">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">analyse_family_ortholog</code><span class="sig-paren">(</span><em>family_id</em>, <em>family</em>, <em>nucleotide</em>, <em>tmp='./'</em>, <em>muscle='muscle'</em>, <em>codeml='codeml'</em>, <em>preserve=False</em>, <em>times=1</em>, <em>min_length=100</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.analyse_family_ortholog" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper function for the analysis of one paralog family. Performs alignment with
<code class="xref py py-meth docutils literal"><span class="pre">alignment.Muscle.run_muscle()</span></code> and codeml analysis with <code class="xref py py-meth docutils literal"><span class="pre">codeml.Codeml.run_codeml()</span></code>.
Subsequently also clustering with <a class="reference internal" href="#wgd.ks_distribution._average_linkage_clustering" title="wgd.ks_distribution._average_linkage_clustering"><code class="xref py py-func docutils literal"><span class="pre">_average_linkage_clustering()</span></code></a> is performed and weighted Ks, Kn and w
values are calculated using <a class="reference internal" href="#wgd.ks_distribution._calculate_weighted_ks" title="wgd.ks_distribution._calculate_weighted_ks"><code class="xref py py-func docutils literal"><span class="pre">_calculate_weighted_ks()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>family_id</strong> &#8211; gene family id</li>
<li><strong>family</strong> &#8211; dictionary with sequences of paralogs</li>
<li><strong>nucleotide</strong> &#8211; nucleotide (CDS) sequences dictionary</li>
<li><strong>tmp</strong> &#8211; tmp directory</li>
<li><strong>muscle</strong> &#8211; muscle path</li>
<li><strong>codeml</strong> &#8211; codeml path</li>
<li><strong>preserve</strong> &#8211; preserve intermediate files</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">csv</span></code> file with results for the paralog family of interest</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.filter_species">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">filter_species</code><span class="sig-paren">(</span><em>results</em>, <em>s1</em>, <em>s2</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.filter_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter out one vs one orthologs from codeml results for two species</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>results</strong> &#8211; </li>
<li><strong>s1</strong> &#8211; </li>
<li><strong>s2</strong> &#8211; </li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.ks_analysis_paranome">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">ks_analysis_paranome</code><span class="sig-paren">(</span><em>nucleotide_sequences</em>, <em>protein_sequences</em>, <em>paralogs</em>, <em>tmp_dir='./tmp'</em>, <em>output_dir='./ks.out'</em>, <em>muscle_path='muscle'</em>, <em>codeml_path='codeml'</em>, <em>check=True</em>, <em>preserve=True</em>, <em>times=1</em>, <em>ignore_prefixes=False</em>, <em>n_cores=4</em>, <em>async=False</em>, <em>min_length=100</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.ks_analysis_paranome" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate a Ks distribution for a whole paranome. Asyncio version</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>nucleotide_sequences</strong> &#8211; CDS fasta file</li>
<li><strong>protein_sequences</strong> &#8211; protein fasta file</li>
<li><strong>paralogs</strong> &#8211; file with paralog families</li>
<li><strong>tmp_dir</strong> &#8211; tmp directory</li>
<li><strong>output_dir</strong> &#8211; output directory</li>
<li><strong>muscle_path</strong> &#8211; path to muscle executable</li>
<li><strong>codeml_path</strong> &#8211; path to codeml executable</li>
<li><strong>check</strong> &#8211; check directories before proceeding (safer)</li>
<li><strong>preserve</strong> &#8211; preserve intermediate results (muscle, codeml)</li>
<li><strong>times</strong> &#8211; number of times to perform codeml analysis</li>
<li><strong>ignore_prefixes</strong> &#8211; ignore prefixes in paralog/gene family file (e.g. in ath|AT1G45000, ath| will be ignored)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">some files and plots in the chosen output directory</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="alignment.html"
                        title="previous chapter">Alignment tools</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="mix.html"
                        title="next chapter">Mixture modeling</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/ks.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="mix.html" title="Mixture modeling"
             >next</a> |</li>
        <li class="right" >
          <a href="alignment.html" title="Alignment tools"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 0.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Arthur Zwaenepoel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.6.
    </div>
  </body>
</html>