
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Ks distribution module &#8212; wgd 1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Phylogenetic tools module" href="phy.html" />
    <link rel="prev" title="Alignment tools" href="alignment.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phy.html" title="Phylogenetic tools module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="alignment.html" title="Alignment tools"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-wgd.ks_distribution">
<span id="ks-distribution-module"></span><h1>Ks distribution module<a class="headerlink" href="#module-wgd.ks_distribution" title="Permalink to this headline">¶</a></h1>
<p>Arthur Zwaenepoel - 2017</p>
<p>Tools for whole paranome and one-vs.-one ortholog Ks distribution construction.
Implemented in parallelized fashion. Please find the functions for node-based weighting in the <code class="docutils literal"><span class="pre">phy</span></code> module.</p>
<dl class="function">
<dt id="wgd.ks_distribution._calculate_weighted_ks">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">_calculate_weighted_ks</code><span class="sig-paren">(</span><em>clustering</em>, <em>pairwise_estimates</em>, <em>pairwise_distances=None</em>, <em>family_id=None</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution._calculate_weighted_ks" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate weighted Ks, Kn and w values following the procedure as outlined in Vanneste et al. (2013)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>clustering</strong> – clustering results as produced with <a class="reference internal" href="#wgd.ks_distribution._weighting" title="wgd.ks_distribution._weighting"><code class="xref py py-func docutils literal"><span class="pre">_weighting()</span></code></a>, can be by average linkage clustering
or phylogenetic means (phyml, fasttree)</li>
<li><strong>pairwise_estimates</strong> – pairwise Ks, Kn and w estimates as produced by <code class="xref py py-meth docutils literal"><span class="pre">codeml.Codeml.run_codeml()</span></code></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a pandas data frame of weighted Ks, Kn and w values.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution._weighting">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">_weighting</code><span class="sig-paren">(</span><em>pairwise_estimates</em>, <em>msa=None</em>, <em>method='alc'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution._weighting" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper for different weighting methods.
The fastest method is average linkage clustering based on Ks values,
Other methods use phylogenetic trees (phyml or fasttree)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pairwise_estimates</strong> – results dictionary from <code class="xref py py-meth docutils literal"><span class="pre">codeml.Codeml.run()</span></code> output</li>
<li><strong>msa</strong> – protein multiple sequence alignment file path (optional, not necessary for method=``alc``)</li>
<li><strong>method</strong> – method, one of <code class="docutils literal"><span class="pre">alc|phyml|fasttree</span></code></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">clustering data structure and pairwise leaf distances (not if method=``alc``)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.add_alignment_stats">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">add_alignment_stats</code><span class="sig-paren">(</span><em>df</em>, <em>stats</em>, <em>l</em>, <em>l_stripped</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.add_alignment_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Add alignment statistics to the data frame</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>df</strong> – pandas data frame</li>
<li><strong>stats</strong> – stats dict see <code class="xref py py-func docutils literal"><span class="pre">alignment.pairwise_alignment_stats()</span></code></li>
<li><strong>l</strong> – alignment length</li>
<li><strong>l_stripped</strong> – stripped alignment length</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">data frame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.analyse_family">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">analyse_family</code><span class="sig-paren">(</span><em>family_id</em>, <em>family</em>, <em>nucleotide</em>, <em>tmp='./'</em>, <em>muscle='muscle'</em>, <em>codeml='codeml'</em>, <em>prank='prank'</em>, <em>preserve=False</em>, <em>times=1</em>, <em>min_length=100</em>, <em>method='alc'</em>, <em>aligner='muscle'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.analyse_family" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper function for the analysis of one paralog family. Performs alignment with
<code class="xref py py-meth docutils literal"><span class="pre">alignment.MSA.run_aligner()</span></code> and codeml analysis with <code class="xref py py-meth docutils literal"><span class="pre">codeml.Codeml.run_codeml()</span></code>.
Subsequently also clustering with <code class="xref py py-func docutils literal"><span class="pre">_average_linkage_clustering()</span></code> is performed and weighted Ks, Kn and w
values are calculated using <a class="reference internal" href="#wgd.ks_distribution._calculate_weighted_ks" title="wgd.ks_distribution._calculate_weighted_ks"><code class="xref py py-func docutils literal"><span class="pre">_calculate_weighted_ks()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>family_id</strong> – gene family id</li>
<li><strong>family</strong> – dictionary with sequences of paralogs</li>
<li><strong>nucleotide</strong> – nucleotide (CDS) sequences dictionary</li>
<li><strong>tmp</strong> – tmp directory</li>
<li><strong>muscle</strong> – muscle path</li>
<li><strong>codeml</strong> – codeml path</li>
<li><strong>preserve</strong> – preserve intermediate files</li>
<li><strong>times</strong> – number of times to perform ML estimation of Ks, Ka and omega values</li>
<li><strong>min_length</strong> – minimum length of the stripped multiple sequence alignment</li>
<li><strong>method</strong> – weighting method, from fast to slow: <code class="docutils literal"><span class="pre">alc</span></code>, <code class="docutils literal"><span class="pre">fasttree</span></code>, <code class="docutils literal"><span class="pre">phyml</span></code></li>
<li><strong>aligner</strong> – alignment program</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><code class="docutils literal"><span class="pre">csv</span></code> file with results for the paralog family of interest</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.ks_analysis_one_vs_one">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">ks_analysis_one_vs_one</code><span class="sig-paren">(</span><em>nucleotide_sequences</em>, <em>protein_sequences</em>, <em>gene_families</em>, <em>tmp_dir='./tmp'</em>, <em>output_dir='./ks.out'</em>, <em>muscle_path='muscle'</em>, <em>codeml_path='codeml'</em>, <em>prank_path='prank'</em>, <em>aligner='muscle'</em>, <em>preserve=True</em>, <em>times=1</em>, <em>n_cores=4</em>, <em>async=False</em>, <em>min_length=100</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.ks_analysis_one_vs_one" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate a Ks distribution for one vs. one orthologs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>nucleotide_sequences</strong> – sequence dictionary</li>
<li><strong>protein_sequences</strong> – protein sequence dictionary</li>
<li><strong>paralogs</strong> – file with paralog families</li>
<li><strong>tmp_dir</strong> – tmp directory</li>
<li><strong>output_dir</strong> – output directory</li>
<li><strong>muscle_path</strong> – path to muscle executable</li>
<li><strong>codeml_path</strong> – path to codeml executable</li>
<li><strong>prank_path</strong> – path to prank executable</li>
<li><strong>preserve</strong> – preserve intermediate results (muscle, codeml)</li>
<li><strong>times</strong> – number of times to perform codeml analysis</li>
<li><strong>ignore_prefixes</strong> – ignore prefixes in paralog/gene family file (e.g. in ath|AT1G45000, ath| will be ignored)</li>
<li><strong>async</strong> – use asyncio library for parallelization</li>
<li><strong>min_length</strong> – minimum MSA length</li>
<li><strong>aligner</strong> – aligner to use (muslce|prank)</li>
<li><strong>n_cores</strong> – number of CPU cores to use</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">data frame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.ks_distribution.ks_analysis_paranome">
<code class="descclassname">wgd.ks_distribution.</code><code class="descname">ks_analysis_paranome</code><span class="sig-paren">(</span><em>nucleotide_sequences</em>, <em>protein_sequences</em>, <em>paralogs</em>, <em>tmp_dir='./tmp'</em>, <em>output_dir='./ks.out'</em>, <em>muscle_path='muscle'</em>, <em>codeml_path='codeml'</em>, <em>prank_path='prank'</em>, <em>preserve=True</em>, <em>times=1</em>, <em>ignore_prefixes=False</em>, <em>n_cores=4</em>, <em>async=False</em>, <em>min_length=100</em>, <em>method='alc'</em>, <em>aligner='muscle'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.ks_distribution.ks_analysis_paranome" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate a Ks distribution for a whole paranome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>nucleotide_sequences</strong> – sequence dictionary</li>
<li><strong>protein_sequences</strong> – protein sequence dictionary</li>
<li><strong>paralogs</strong> – file with paralog families</li>
<li><strong>tmp_dir</strong> – tmp directory</li>
<li><strong>output_dir</strong> – output directory</li>
<li><strong>muscle_path</strong> – path to muscle executable</li>
<li><strong>codeml_path</strong> – path to codeml executable</li>
<li><strong>preserve</strong> – preserve intermediate results (muscle, codeml)</li>
<li><strong>times</strong> – number of times to perform codeml analysis</li>
<li><strong>ignore_prefixes</strong> – ignore prefixes in paralog/gene family file (e.g. in ath|AT1G45000, ath| will be ignored)</li>
<li><strong>async</strong> – use asyncio library for parallelization</li>
<li><strong>min_length</strong> – minimum MSA length</li>
<li><strong>method</strong> – method to use, from fast to slow: <code class="docutils literal"><span class="pre">alc</span></code>, <code class="docutils literal"><span class="pre">fasttree</span></code>, <code class="docutils literal"><span class="pre">phyml</span></code></li>
<li><strong>aligner</strong> – alignment program to use (muscle|prank)</li>
<li><strong>n_cores</strong> – number of CPU cores to use</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">data frame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="alignment.html"
                        title="previous chapter">Alignment tools</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="phy.html"
                        title="next chapter">Phylogenetic tools module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/ks.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phy.html" title="Phylogenetic tools module"
             >next</a> |</li>
        <li class="right" >
          <a href="alignment.html" title="Alignment tools"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Arthur Zwaenepoel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>