
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Co-linearity analyses module &#8212; wgd 1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Various utilities" href="utils.html" />
    <link rel="prev" title="Mixture modeling" href="mix.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="utils.html" title="Various utilities"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="mix.html" title="Mixture modeling"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-wgd.colinearity">
<span id="co-linearity-analyses-module"></span><h1>Co-linearity analyses module<a class="headerlink" href="#module-wgd.colinearity" title="Permalink to this headline">¶</a></h1>
<p>Arthur Zwaenepoel - 2017</p>
<p>The co-linearity analysis functions currently only support intragenomic analyses.</p>
<dl class="function">
<dt id="wgd.colinearity.get_anchor_pairs">
<code class="descclassname">wgd.colinearity.</code><code class="descname">get_anchor_pairs</code><span class="sig-paren">(</span><em>anchors_file</em>, <em>ks_distribution=None</em>, <em>out_file='anchors_ks.csv'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.colinearity.get_anchor_pairs" title="Permalink to this definition">¶</a></dt>
<dd><p>Get anchor pairs and their corresponding Ks values (if provided)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>anchors_file</strong> – anchorpoints.txt output from I-ADHoRe 3.0</li>
<li><strong>ks_distribution</strong> – Ks distribution dataf rame</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">pandas dataframe(s): anchors and data frame</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.colinearity.run_adhore">
<code class="descclassname">wgd.colinearity.</code><code class="descname">run_adhore</code><span class="sig-paren">(</span><em>config_file</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.colinearity.run_adhore" title="Permalink to this definition">¶</a></dt>
<dd><p>Run I-ADHoRe for a given config file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>config_file</strong> – path to I-ADHoRe configuration file</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.colinearity.segments_to_chords_table">
<code class="descclassname">wgd.colinearity.</code><code class="descname">segments_to_chords_table</code><span class="sig-paren">(</span><em>segments_file</em>, <em>genome</em>, <em>output_file='chords.tsv'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.colinearity.segments_to_chords_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Create chords table for visualization in a chord diagram. Uses the segments.txt
output of I-ADHoRe. Chords are defined by a source chromosome and a target 
chromosome and begin and end coordinates for each chromosome respectively.</p>
<p>TODO: the length of each syntenic block should be included in the table as well
with length defined as number of genes, not physical length.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>segments_file</strong> – pat to the I-ADHoRe segments.txt output file</li>
<li><strong>genome</strong> – a <code class="xref py py-func docutils literal"><span class="pre">gff_parser.Genome</span> <span class="pre">object()</span></code></li>
<li><strong>output_file</strong> – output file name</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.colinearity.write_config_adhore">
<code class="descclassname">wgd.colinearity.</code><code class="descname">write_config_adhore</code><span class="sig-paren">(</span><em>gene_lists</em>, <em>families</em>, <em>config_file_name='i-adhore.conf'</em>, <em>genome='genome'</em>, <em>output_path='i-adhore_out'</em>, <em>gap_size=30</em>, <em>cluster_gap=35</em>, <em>q_value=0.75</em>, <em>prob_cutoff=0.01</em>, <em>anchor_points=3</em>, <em>alignment_method='gg2'</em>, <em>level_2_only='false'</em>, <em>table_type='family'</em>, <em>multiple_hypothesis_correction='FDR'</em>, <em>visualize_ghm='false'</em>, <em>visualize_alignment='true'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.colinearity.write_config_adhore" title="Permalink to this definition">¶</a></dt>
<dd><p>Write out the config file for I-ADHoRe. See I-ADHoRe manual for information on
parameter settings.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gene_lists</strong> – directory with gene lists per chromosome</li>
<li><strong>families</strong> – file with gene to family mapping</li>
<li><strong>config_file_name</strong> – name for the config file</li>
<li><strong>genome</strong> – genome name</li>
<li><strong>output_path</strong> – output path name</li>
<li><strong>gap_size</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>cluster_gap</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>q_value</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>prob_cutoff</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>anchor_points</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>alignment_method</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>level_2_only</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>table_type</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>multiple_hypothesis_correction</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>visualize_ghm</strong> – see I-ADHoRe 3.0 documentation</li>
<li><strong>visualize_alignment</strong> – see I-ADHoRe 3.0 documentation</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">configuration file see I-ADHoRe 3.0 documentation</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.colinearity.write_families_file">
<code class="descclassname">wgd.colinearity.</code><code class="descname">write_families_file</code><span class="sig-paren">(</span><em>families</em>, <em>all_genes</em>, <em>output_file='families.tsv'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.colinearity.write_families_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Write out families file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>families</strong> – gene families</li>
<li><strong>all_genes</strong> – set object with all genes for the species of interest</li>
<li><strong>output_file</strong> – output file name</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">nada</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.colinearity.write_gene_lists">
<code class="descclassname">wgd.colinearity.</code><code class="descname">write_gene_lists</code><span class="sig-paren">(</span><em>genome</em>, <em>output_dir='gene_lists'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.colinearity.write_gene_lists" title="Permalink to this definition">¶</a></dt>
<dd><p>Write out the gene lists</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>genome</strong> – Genome object</li>
<li><strong>output_dir</strong> – output directory</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="mix.html"
                        title="previous chapter">Mixture modeling</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="utils.html"
                        title="next chapter">Various utilities</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/syn.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="utils.html" title="Various utilities"
             >next</a> |</li>
        <li class="right" >
          <a href="mix.html" title="Mixture modeling"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Arthur Zwaenepoel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>