<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Various utilities &mdash; wgd 1.0.0 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="wgd 1.0.0 documentation" href="index.html" />
    <link rel="next" title="Visualization module" href="viz.html" />
    <link rel="prev" title="Co-linearity analyses module" href="syn.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="viz.html" title="Visualization module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="syn.html" title="Co-linearity analyses module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="various-utilities">
<h1>Various utilities<a class="headerlink" href="#various-utilities" title="Permalink to this headline">¶</a></h1>
<p>Probably not of specific interest.</p>
<span class="target" id="module-wgd.utils"></span><p>Arthur Zwaenepoel - 2017</p>
<dl class="class">
<dt id="wgd.utils.Genome">
<em class="property">class </em><code class="descclassname">wgd.utils.</code><code class="descname">Genome</code><a class="headerlink" href="#wgd.utils.Genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Class that represents a structural annotation.
Collects several nice data structures for a genome and parsers for various
genomic data file formats (e.g. gff, fasta, ...)</p>
<dl class="method">
<dt id="wgd.utils.Genome.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.Genome.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Genome.genome: dictionary with a full representation
Genome.gene_lists: dictionary with ordered lists of genes per
chromosome (as for I-ADHoRe)</p>
</dd></dl>

<dl class="method">
<dt id="wgd.utils.Genome.karyotype_json">
<code class="descname">karyotype_json</code><span class="sig-paren">(</span><em>out_file='genome.json'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.Genome.karyotype_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate karyotype data file in json format (as per Circos.js/d3.js)</p>
</dd></dl>

<dl class="method">
<dt id="wgd.utils.Genome.parse_plaza_gff">
<code class="descname">parse_plaza_gff</code><span class="sig-paren">(</span><em>gff_file</em>, <em>keyword='mRNA'</em>, <em>id_string='Parent'</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.Genome.parse_plaza_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a PLAZA annotation file into a genome dictionary</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gff_file</strong> &#8211; input gff (PLAZA style)</li>
<li><strong>keyword</strong> &#8211; keyword for elements to parse out</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="wgd.utils._random_color">
<code class="descclassname">wgd.utils.</code><code class="descname">_random_color</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils._random_color" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a random hex color</p>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.check_dirs">
<code class="descclassname">wgd.utils.</code><code class="descname">check_dirs</code><span class="sig-paren">(</span><em>tmp_dir</em>, <em>output_dir</em>, <em>prompt</em>, <em>preserve</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.check_dirs" title="Permalink to this definition">¶</a></dt>
<dd><p>Check directories needed
:param tmp_dir: tmp directory
:param output_dir: output directory
:param prompt: prompt for overwrites (boolean)?
:param preserve: preserve MSA files (boolean)?
:return: nothing</p>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.filter_one_vs_one_families">
<code class="descclassname">wgd.utils.</code><code class="descname">filter_one_vs_one_families</code><span class="sig-paren">(</span><em>gene_families</em>, <em>s1</em>, <em>s2</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.filter_one_vs_one_families" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter one-vs-one ortholog containing families for two given species.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gene_families</strong> &#8211; </li>
<li><strong>s1</strong> &#8211; </li>
<li><strong>s2</strong> &#8211; </li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="wgd.utils.gaussian_kde">
<em class="property">class </em><code class="descclassname">wgd.utils.</code><code class="descname">gaussian_kde</code><span class="sig-paren">(</span><em>dataset</em>, <em>bw_method=None</em>, <em>weights=None</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.gaussian_kde" title="Permalink to this definition">¶</a></dt>
<dd><p>from: <a class="reference external" href="https://stackoverflow.com/questions/27623919/weighted-gaussian-kernel-density-estimation-in-python">https://stackoverflow.com/questions/27623919/weighted-gaussian-kernel-density-estimation-in-python</a></p>
<p>Representation of a kernel-density estimate using Gaussian kernels.</p>
<p>Kernel density estimation is a way to estimate the probability density
function (PDF) of a random variable in a non-parametric way.
<cite>gaussian_kde</cite> works for both uni-variate and multi-variate data.   It
includes automatic bandwidth determination.  The estimation works best for
a unimodal distribution; bimodal or multi-modal distributions tend to be
oversmoothed.</p>
<dl class="docutils">
<dt>dataset <span class="classifier-delimiter">:</span> <span class="classifier">array_like</span></dt>
<dd>Datapoints to estimate from. In case of univariate data this is a 1-D
array, otherwise a 2-D array with shape (# of dims, # of data).</dd>
<dt>bw_method <span class="classifier-delimiter">:</span> <span class="classifier">str, scalar or callable, optional</span></dt>
<dd>The method used to calculate the estimator bandwidth.  This can be
&#8216;scott&#8217;, &#8216;silverman&#8217;, a scalar constant or a callable.  If a scalar,
this will be used directly as <cite>kde.factor</cite>.  If a callable, it should
take a <cite>gaussian_kde</cite> instance as only parameter and return a scalar.
If None (default), &#8216;scott&#8217; is used.  See Notes for more details.</dd>
<dt>weights <span class="classifier-delimiter">:</span> <span class="classifier">array_like, shape (n, ), optional, default: None</span></dt>
<dd>An array of weights, of the same shape as <cite>x</cite>.  Each value in <cite>x</cite>
only contributes its associated weight towards the bin count
(instead of 1).</dd>
</dl>
<dl class="docutils">
<dt>dataset <span class="classifier-delimiter">:</span> <span class="classifier">ndarray</span></dt>
<dd>The dataset with which <cite>gaussian_kde</cite> was initialized.</dd>
<dt>d <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Number of dimensions.</dd>
<dt>n <span class="classifier-delimiter">:</span> <span class="classifier">int</span></dt>
<dd>Number of datapoints.</dd>
<dt>neff <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd>Effective sample size using Kish&#8217;s approximation.</dd>
<dt>factor <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd>The bandwidth factor, obtained from <cite>kde.covariance_factor</cite>, with which
the covariance matrix is multiplied.</dd>
<dt>covariance <span class="classifier-delimiter">:</span> <span class="classifier">ndarray</span></dt>
<dd>The covariance matrix of <cite>dataset</cite>, scaled by the calculated bandwidth
(<cite>kde.factor</cite>).</dd>
<dt>inv_cov <span class="classifier-delimiter">:</span> <span class="classifier">ndarray</span></dt>
<dd>The inverse of <cite>covariance</cite>.</dd>
</dl>
<dl class="docutils">
<dt>kde.evaluate(points) <span class="classifier-delimiter">:</span> <span class="classifier">ndarray</span></dt>
<dd>Evaluate the estimated pdf on a provided set of points.</dd>
<dt>kde(points) <span class="classifier-delimiter">:</span> <span class="classifier">ndarray</span></dt>
<dd>Same as kde.evaluate(points)</dd>
<dt>kde.pdf(points) <span class="classifier-delimiter">:</span> <span class="classifier">ndarray</span></dt>
<dd>Alias for <code class="docutils literal"><span class="pre">kde.evaluate(points)</span></code>.</dd>
<dt>kde.set_bandwidth(bw_method=&#8217;scott&#8217;) <span class="classifier-delimiter">:</span> <span class="classifier">None</span></dt>
<dd>Computes the bandwidth, i.e. the coefficient that multiplies the data
covariance matrix to obtain the kernel covariance matrix.
.. versionadded:: 0.11.0</dd>
<dt>kde.covariance_factor <span class="classifier-delimiter">:</span> <span class="classifier">float</span></dt>
<dd>Computes the coefficient (<cite>kde.factor</cite>) that multiplies the data
covariance matrix to obtain the kernel covariance matrix.
The default is <cite>scotts_factor</cite>.  A subclass can overwrite this method
to provide a different method, or set it through a call to
<cite>kde.set_bandwidth</cite>.</dd>
</dl>
<p>Bandwidth selection strongly influences the estimate obtained from the KDE
(much more so than the actual shape of the kernel).  Bandwidth selection
can be done by a &#8220;rule of thumb&#8221;, by cross-validation, by &#8220;plug-in
methods&#8221; or by other means; see <a class="footnote-reference" href="#id10" id="id1">[3]</a>, <a class="footnote-reference" href="#id11" id="id2">[4]</a> for reviews.  <cite>gaussian_kde</cite>
uses a rule of thumb, the default is Scott&#8217;s Rule.</p>
<p>Scott&#8217;s Rule <a class="footnote-reference" href="#id8" id="id3">[1]</a>, implemented as <cite>scotts_factor</cite>, is:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="n">n</span><span class="o">**</span><span class="p">(</span><span class="o">-</span><span class="mf">1.</span><span class="o">/</span><span class="p">(</span><span class="n">d</span><span class="o">+</span><span class="mi">4</span><span class="p">)),</span>
</pre></div>
</div>
<p>with <code class="docutils literal"><span class="pre">n</span></code> the number of data points and <code class="docutils literal"><span class="pre">d</span></code> the number of dimensions.
Silverman&#8217;s Rule <a class="footnote-reference" href="#id9" id="id4">[2]</a>, implemented as <cite>silverman_factor</cite>, is:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>(n * (d + 2) / 4.)**(-1. / (d + 4)).
</pre></div>
</div>
<p>Good general descriptions of kernel density estimation can be found in <a class="footnote-reference" href="#id8" id="id5">[1]</a>
and <a class="footnote-reference" href="#id9" id="id6">[2]</a>, the mathematics for this multi-dimensional implementation can be
found in <a class="footnote-reference" href="#id8" id="id7">[1]</a>.</p>
<table class="docutils footnote" frame="void" id="id8" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td><em>(<a class="fn-backref" href="#id3">1</a>, <a class="fn-backref" href="#id5">2</a>, <a class="fn-backref" href="#id7">3</a>)</em> D.W. Scott, &#8220;Multivariate Density Estimation: Theory, Practice, and
Visualization&#8221;, John Wiley &amp; Sons, New York, Chicester, 1992.</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id9" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[2]</td><td><em>(<a class="fn-backref" href="#id4">1</a>, <a class="fn-backref" href="#id6">2</a>)</em> B.W. Silverman, &#8220;Density Estimation for Statistics and Data
Analysis&#8221;, Vol. 26, Monographs on Statistics and Applied Probability,
Chapman and Hall, London, 1986.</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id10" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[3]</a></td><td>B.A. Turlach, &#8220;Bandwidth Selection in Kernel Density Estimation: A
Review&#8221;, CORE and Institut de Statistique, Vol. 19, pp. 1-33, 1993.</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id11" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[4]</a></td><td>D.M. Bashtannyk and R.J. Hyndman, &#8220;Bandwidth selection for kernel
conditional density estimation&#8221;, Computational Statistics &amp; Data
Analysis, Vol. 36, pp. 279-298, 2001.</td></tr>
</tbody>
</table>
<p>Generate some random two-dimensional data:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">measure</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="s2">&quot;Measurement model, return two coupled measurements.&quot;</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">m1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">m2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">n</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="k">return</span> <span class="n">m1</span><span class="o">+</span><span class="n">m2</span><span class="p">,</span> <span class="n">m1</span><span class="o">-</span><span class="n">m2</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">m1</span><span class="p">,</span> <span class="n">m2</span> <span class="o">=</span> <span class="n">measure</span><span class="p">(</span><span class="mi">2000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">xmin</span> <span class="o">=</span> <span class="n">m1</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">xmax</span> <span class="o">=</span> <span class="n">m1</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ymin</span> <span class="o">=</span> <span class="n">m2</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ymax</span> <span class="o">=</span> <span class="n">m2</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
</pre></div>
</div>
<p>Perform a kernel density estimate on the data:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mgrid</span><span class="p">[</span><span class="n">xmin</span><span class="p">:</span><span class="n">xmax</span><span class="p">:</span><span class="mi">100j</span><span class="p">,</span> <span class="n">ymin</span><span class="p">:</span><span class="n">ymax</span><span class="p">:</span><span class="mi">100j</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">X</span><span class="o">.</span><span class="n">ravel</span><span class="p">(),</span> <span class="n">Y</span><span class="o">.</span><span class="n">ravel</span><span class="p">()])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">m1</span><span class="p">,</span> <span class="n">m2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kernel</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">gaussian_kde</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">kernel</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">X</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
<p>Plot the results:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">rot90</span><span class="p">(</span><span class="n">Z</span><span class="p">),</span> <span class="n">cmap</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">gist_earth_r</span><span class="p">,</span>
<span class="gp">... </span>          <span class="n">extent</span><span class="o">=</span><span class="p">[</span><span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">m1</span><span class="p">,</span> <span class="n">m2</span><span class="p">,</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="n">xmin</span><span class="p">,</span> <span class="n">xmax</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="n">ymin</span><span class="p">,</span> <span class="n">ymax</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<dl class="method">
<dt id="wgd.utils.gaussian_kde._compute_covariance">
<code class="descname">_compute_covariance</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.gaussian_kde._compute_covariance" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes the covariance matrix for each Gaussian kernel using
covariance_factor().</p>
</dd></dl>

<dl class="method">
<dt id="wgd.utils.gaussian_kde.evaluate">
<code class="descname">evaluate</code><span class="sig-paren">(</span><em>points</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.gaussian_kde.evaluate" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate the estimated pdf on a set of points.</p>
<dl class="docutils">
<dt>points <span class="classifier-delimiter">:</span> <span class="classifier">(# of dimensions, # of points)-array</span></dt>
<dd>Alternatively, a (# of dimensions,) vector can be passed in and
treated as a single point.</dd>
</dl>
<dl class="docutils">
<dt>values <span class="classifier-delimiter">:</span> <span class="classifier">(# of points,)-array</span></dt>
<dd>The values at each point.</dd>
</dl>
<dl class="docutils">
<dt>ValueError <span class="classifier-delimiter">:</span> <span class="classifier">if the dimensionality of the input points is different than</span></dt>
<dd>the dimensionality of the KDE.</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="wgd.utils.gaussian_kde.set_bandwidth">
<code class="descname">set_bandwidth</code><span class="sig-paren">(</span><em>bw_method=None</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.gaussian_kde.set_bandwidth" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the estimator bandwidth with given method.</p>
<p>The new bandwidth calculated after a call to <cite>set_bandwidth</cite> is used
for subsequent evaluations of the estimated density.</p>
<dl class="docutils">
<dt>bw_method <span class="classifier-delimiter">:</span> <span class="classifier">str, scalar or callable, optional</span></dt>
<dd>The method used to calculate the estimator bandwidth.  This can be
&#8216;scott&#8217;, &#8216;silverman&#8217;, a scalar constant or a callable.  If a
scalar, this will be used directly as <cite>kde.factor</cite>.  If a callable,
it should take a <cite>gaussian_kde</cite> instance as only parameter and
return a scalar.  If None (default), nothing happens; the current
<cite>kde.covariance_factor</cite> method is kept.</dd>
</dl>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.11.</span></p>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="o">-</span><span class="mi">7</span><span class="p">,</span> <span class="o">-</span><span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mf">5.</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kde</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">gaussian_kde</span><span class="p">(</span><span class="n">x1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">xs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y1</span> <span class="o">=</span> <span class="n">kde</span><span class="p">(</span><span class="n">xs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kde</span><span class="o">.</span><span class="n">set_bandwidth</span><span class="p">(</span><span class="n">bw_method</span><span class="o">=</span><span class="s1">&#39;silverman&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y2</span> <span class="o">=</span> <span class="n">kde</span><span class="p">(</span><span class="n">xs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kde</span><span class="o">.</span><span class="n">set_bandwidth</span><span class="p">(</span><span class="n">bw_method</span><span class="o">=</span><span class="n">kde</span><span class="o">.</span><span class="n">factor</span> <span class="o">/</span> <span class="mf">3.</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y3</span> <span class="o">=</span> <span class="n">kde</span><span class="p">(</span><span class="n">xs</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_subplot</span><span class="p">(</span><span class="mi">111</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x1</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">x1</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="mf">4.</span> <span class="o">*</span> <span class="n">x1</span><span class="o">.</span><span class="n">size</span><span class="p">),</span> <span class="s1">&#39;bo&#39;</span><span class="p">,</span>
<span class="gp">... </span>        <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Data points (rescaled)&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">xs</span><span class="p">,</span> <span class="n">y1</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Scott (default)&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">xs</span><span class="p">,</span> <span class="n">y2</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Silverman&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">xs</span><span class="p">,</span> <span class="n">y3</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Const (1/3 * Silverman)&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="wgd.utils.get_gfs_for_species">
<code class="descclassname">wgd.utils.</code><code class="descname">get_gfs_for_species</code><span class="sig-paren">(</span><em>gene_family_dict</em>, <em>gene_pattern</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.get_gfs_for_species" title="Permalink to this definition">¶</a></dt>
<dd><p>Get non-singleton gene families for a species of interest
:param gene_family_dict: gene family dictionary
:param species: species of interest
:return: dictionairy with gene families</p>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.get_number_of_sp">
<code class="descclassname">wgd.utils.</code><code class="descname">get_number_of_sp</code><span class="sig-paren">(</span><em>genes</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.get_number_of_sp" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the number of unique species in a list of gene IDs.
Will be approximate since it is based on the assumption that
the leading non-digit part identifies the species
(which is not always the case e.g. mitochondrial genes etc.)</p>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.get_paralogs_fasta">
<code class="descclassname">wgd.utils.</code><code class="descname">get_paralogs_fasta</code><span class="sig-paren">(</span><em>input_fasta</em>, <em>selected_paralogs</em>, <em>output_fasta</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.get_paralogs_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the fasta file associated with the paralogs in a slice from a Ks distribution data frame</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>input_fasta</strong> &#8211; fasta file</li>
<li><strong>selected_paralogs</strong> &#8211; Data frame (slice)</li>
<li><strong>output_fasta</strong> &#8211; output fasta file</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">nada</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.get_sequences">
<code class="descclassname">wgd.utils.</code><code class="descname">get_sequences</code><span class="sig-paren">(</span><em>paralog_dict</em>, <em>sequences</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.get_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch sequences from a fasta file or sequence dict and put them in a two level dictionairy
{gene_family: {gene: seq, gene: seq, ...}, ...}
:param paralog_dict:
:return: two-level dictionairy</p>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.process_gene_families">
<code class="descclassname">wgd.utils.</code><code class="descname">process_gene_families</code><span class="sig-paren">(</span><em>gene_family_file</em>, <em>ignore_prefix=False</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.process_gene_families" title="Permalink to this definition">¶</a></dt>
<dd><p>Processes a raw gene family file as e.g. from OrthoMCL into a generic dictionary structure
OrthoMCL raw file consists of one gene family per line, including tab separated gene IDs,
without gene family ID.</p>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.read_fasta">
<code class="descclassname">wgd.utils.</code><code class="descname">read_fasta</code><span class="sig-paren">(</span><em>fasta_file</em>, <em>prefix=None</em>, <em>split_on_pipe=True</em>, <em>split_on_whitespace=True</em>, <em>raw=False</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.read_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Generic fastafile reader
Returns a dictionairy {ID: sequence, ID: sequence, ...}.</p>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.translate_cds">
<code class="descclassname">wgd.utils.</code><code class="descname">translate_cds</code><span class="sig-paren">(</span><em>sequence_dict</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.translate_cds" title="Permalink to this definition">¶</a></dt>
<dd><p>Just another CDS to protein translater</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sequence_dict</strong> &#8211; dictionary with gene IDs and CDS sequences</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">dictionary with gene IDs and proteins sequences</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="wgd.utils.write_fasta">
<code class="descclassname">wgd.utils.</code><code class="descname">write_fasta</code><span class="sig-paren">(</span><em>seq_dict</em>, <em>output_file</em><span class="sig-paren">)</span><a class="headerlink" href="#wgd.utils.write_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Write a sequence dictionary to a fasta file</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>seq_dict</strong> &#8211; </li>
<li><strong>output_file</strong> &#8211; </li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="syn.html"
                        title="previous chapter">Co-linearity analyses module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="viz.html"
                        title="next chapter">Visualization module</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/utils.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="viz.html" title="Visualization module"
             >next</a> |</li>
        <li class="right" >
          <a href="syn.html" title="Co-linearity analyses module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">wgd 1.0.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2017, Arthur Zwaenepoel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.6.
    </div>
  </body>
</html>