# Simple command line tools for $Ks$-based detection of ancient whole genome duplications

## Abstract

In the past decades, whole genome duplications have been uncovered in almost
all major lineages of life on earth. Today, uncovering such events has become a
standard element of common genome analysis workflows. One of the common
approaches to find evidence for ancient WGDs in a genome is to construct a
$K_S$ distribution, recording the estimated synonymous substitution rates
($K_S$) for the full paranome of a species of interest. Despite the increased
interest in WGDs and the acknowledgement of their evolutionary importance,
common analysis tools seem to be relatively scarce. Here we present an easy to
use command-line tool for $K_S$ distribution construction named `wgd`. The
`wgd` program provides the commonly used $K_S$ analysis workflows in a
convenient manner for bioinformaticians, omitting the need for custom scripts
or online submissions.  `wgd` is available at: [gitlab].

## Introduction

In 1970, the japanese geneticist Susumu Ohno hypothesized that besides gene
duplication, whole genome duplication (WGD) had probably been of considerable
importance in the evolution of life on earth (\citet{Ohno1970}, Chapter XVI).
With the advent of the genomic age, Ohno's profetic claims have become
increasingly supported by genomic data. Indeed, a considerable amount of
ancient WGDs have been uncovered in plants, fungi and vertebrates.

One of the main approaches, and perhaps the most traditional one, for revealing
ancient WGDs using genomic data is the construction of whole paranome $K_S$ 
distributions. This approach was used by Wolfe & Shields (check) 





